### PR TITLE
AES-GCM streaming: fix IV caching

### PR DESCRIPTION
**Pull request by SparkiDev**
# Description

AES-GCM stremaing was caching IV even when larger than buffer copying into.
Instead, require calls to wc_AesGcmSetIV() or wc_AesGcmSetExtIV() to cache IV.
wc_AesGcmInit() now uses passed in IV or retrieves from cache.

Fixes zd#15423

# Testing

Standard testing.

# Checklist

 - [ ] added tests
 - [ ] updated/added doxygen
 - [ ] updated appropriate READMEs
 - [ ] Updated manual and documentation

### CODE REVIEW

File: wolfcrypt/src/aes.c [hunk 1]
```diff
@@ -9523,6 +9537,9 @@ int wc_AesGcmEncryptFinal(Aes* aes, byte* authTag, word32 authTagSz)
  *
  * Must call wc_AesInit() before calling this function.
  *
+ * Call wc_AesGcmSetExtIV() before calling this function to use FIPS external IV
```
Comment (tim-weller-wolfssl) [inline wolfcrypt/src/aes.c:9540]: Is `wc_AesGcmSetExtIV()` only used for FIPS (i.e. without long IVs), or is this API expected to work in non-FIPS scenarios (i.e. with long IVs), as a stand-in for `AesGcmSetIV()`?

File: wolfcrypt/src/aes.c [hunk 2]
```diff
@@ -9366,24 +9363,41 @@ int wc_AesGcmEncryptInit(Aes* aes, const byte* key, word32 len, const byte* iv,
...
  * @param [in, out] aes   AES object.
  * @param [in]      key   Buffer holding key.
  * @param [in]      len   Length of key in bytes.
  * @param [in]      iv    Buffer holding IV/nonce.
  * @param [in]      ivSz  Length of IV/nonce in bytes.
  * @return  0 on success.
- * @return  BAD_FUNC_ARG when aes is NULL, or a length is non-zero but buffer
- *          is NULL, or the IV is NULL and no previous IV has been set.
+ * @return  BAD_FUNC_ARG when aes is NULL, key length is non-zero but key
+ *          is NULL, or the IV is NULL or ivOutSz is not the same as chached
```
Comment (tim-weller-wolfssl) [inline wolfcrypt/src/aes.c:9381]: spelling: cached

**Review by tim-weller-wolfssl — APPROVED** [review]

File: wolfcrypt/src/aes.c (hunk 1 above)
Comment (SparkiDev) [reply wolfcrypt/src/aes.c:9540]: If you look at the implementation of wc_AesGcmSetExtIV() you will see that it copies the IV into aes->reg.
It then sets the invokeCtr as required by SP800-38D which is a NIST standard for FIPS.
This API was added for FIPS.

File: wolfcrypt/src/aes.c (hunk 2 above)
Comment (SparkiDev) [reply wolfcrypt/src/aes.c]: Fixed

**Comment by tim-weller-wolfssl** [comment]
Hi @SparkiDev!  This fix has been verified by the original issue reporter on [zd15423](https://wolfssl.zendesk.com/agent/tickets/15423).

File: wolfcrypt/src/aes.c (hunk 1 above)
Comment (tim-weller-wolfssl) [reply wolfcrypt/src/aes.c:9540]: I see!  Thank-you for the clarification. .:)

**Review by JacobBarthelmeh — APPROVED** [review]

**Pull request merged**